### PR TITLE
add restrict qualifier to inputs in c backend

### PR DESCRIPTION
**Pull request by benmkw**
as discussed on discord, results can roughly be seen in https://godbolt.org/z/Tb9jMxWfx (ideally this would be quantified with perf measurements at some point)

The efficientnet was tested with `-fsanitize=address,undefined` and it ran fine.

I added separate commits for minor changes which I can also remove if they are not wanted.

### CODE REVIEW

**Comment by geohot** [comment]
Tests failing.

**Comment by benmkw** [comment]
I think the failures were because master was at https://github.com/geohot/tinygrad/commit/d3feea302dc3bdd5361f0913050309b69aa65507

rebased on latest master

File: tinygrad/llops/ops_gpu.py [hunk 1]
```diff
@@ -325,7 +325,7 @@ def codegen(self) -> Callable:
 
     # kernel function definition
     function_name = ("re_S" if self.reduceop else "ew_S") + '_'.join([str(x) for x in self.bufs[0].shape if x != 1])
-    buftypes = [f"{'read_only' if i > 0 else 'write_only'} image2d_t" if isinstance(x._buf, CLImage) else (CLProgram.buffer_prefix+self.buftokens[i].decltype()) for i,x in enumerate(self.bufs)]
+    buftypes = [f"{'read_only' if i > 0 else 'write_only'} image2d_t" if isinstance(x._buf, CLImage) else (CLProgram.buffer_prefix+self.buftokens[i].decltype() + "restrict" if CLANG else "") for i,x in enumerate(self.bufs)]
```
Comment (python273) [inline tinygrad/llops/ops_gpu.py:328]: You need to wrap `if` with parenthesis

Comment (benmkw) [reply tinygrad/llops/ops_gpu.py:328]: thanks for the catch + explanation, fixed in a new commit, I can combine them if wanted.

**Pull request merged**